### ISSUE TITLE
All assertion micros are not executed

### ISSUE THREAD

**Issue by hassanfarid** [philsquared/Catch#933]
I have a single test case, and it can process multiple input files.

File1:
{
"configuration":{}
"assertions":[
 {}, {}
]
}

Similarly, we have file 2 as well.

When I test with only one file at a time, it runs 36 assertions for each. When I run both files together, it only run 36 assertions.

All tests passed (36 assertions in 1 test case).

The problem has begun to appear, when I have started using multiple SECTIONS inside the same test case. When I have no or only one SECTION, all assertions were executed for all test cases. With multiple sections in place, this is for some reason not the case.

The test is something like this,

TEST_CASE("name", "[tags]") {
CHECK();

SECTION("name1") {
CHECK();
}
SECTION("name2") {
CHECK();
}
}

**Comment by hassanfarid** [comment]
I realize that keeping SECTION under for loop is the real problem.

I commented out
//SECTIONS("name"){
//}

and now I see all the assertions running.

**Comment by horenmar** [comment]
Sections do not work with for loops ATM. This is actually mostly by design -- it is an accepted consequence of how Catch tracks what sections it has already executed and which are yet to be executed.

IIRC you can work around this by having the section's name change between iterations (thus making it look like a brand new section), but I am not sure.

**Comment by philsquared** [comment]
yes, for now, making the section name unique on each iteration (i.e. incorporating the loop variable somehow) is the workaround.
Longer term you'd use generators instead, but they're still a little way off.

**Comment by philsquared** [comment]
@hassanfarid did this answer your question for now? Are you ok if we close this issue?

**Comment by hassanfarid** [comment]
Sure - Can we link to an issue that will be developed as suggested above?

**Comment by MonsieurNicolas** [comment]
I think I ran into this in a different context.

The simplified version is this (can be done with lambda or any helper function really):
```c++
TEST_CASE("lambda", "lambda")
{
    int a, b;
    a = b = 1;
    auto f = [&](bool v) {
        SECTION("s1")
        {
            REQUIRE(a == b);
        }
    };
    f(true);
    f(false);
```
It's very common for people to reuse as much test code as possible, and here the behavior ends up being some sections getting skipped.

So while a possible fix, is to simply do something like
```c++
TEST_CASE("lambda", "lambda")
{
    int a, b;
    a = b = 1;
   auto f = [&](bool v) {
        // create a section name containing all of f's parameters  
        DYNAMIC_SECTION("f(" << v << ")")
        {
            SECTION("pof")
            {
                REQUIRE(a == b);
            }
        }
    };
    f(true);
    f(false);
```

The problem is that this is a silent failure (less tests are getting executed), and only careful analysis of test code would reveal the problem.

One possible way to deal with this would be to not allow the same section to be evaluated multiple times as a sibling in the call tree.
I am not too familiar with catch's code to know if there is a cheap way to check for this.
